### ISSUE TITLE
Pin generation on open for version aware file system

### ISSUE THREAD

**Issue by emfdavid** [fsspec/gcsfs#601]
[Here is an example](https://gist.github.com/emfdavid/c4f4e0573444944b3161430d24998c07) of the current behavior of version_aware file systems and open files in GCSFS.

I would have expected a version_aware file system to pin the generation of an object while the file is open so that the reads are consistent like they are when the open url specifies the generation.

If this is the desired behavior, I would be happy to take a look and see if I can fix it. Based on your comment in the [previous issue](https://github.com/fsspec/gcsfs/issues/446#issuecomment-1879307967) I think the [python gcs sdk download_as_bytes api](https://cloud.google.com/python/docs/reference/storage/latest/google.cloud.storage.blob.Blob#google_cloud_storage_blob_Blob_download_as_bytes) should support what we need implementing etag matching similar to what exists in S3 but I haven't compared the details yet.

**Comment by martindurant** [comment]
> I would have expected a version_aware file system to pin the generation of an object while the file is open so that the reads are consistent like they are when the open url specifies the generation.

This is a reasonable expectation

**Comment by martindurant** [comment]
FWIW, I think s3fs does this right even without version-aware by saving the etag.

**Comment by emfdavid** [comment]
Is this worth asking Google about?

**Comment by martindurant** [comment]
No, I'm pretty sure we can do this by ourselves with the available information
